rasa/core/lock_store.py

```diff
@@ -0,0 +1,227 @@
+            self.save_lock(lock)
+
+    def get_or_create_lock(self, conversation_id: Text) -> TicketLock:
+        """Fetch existing lock for `conversation_id` or create a new one if
+        it doesn't exist."""
+
+        existing_lock = self.get_lock(conversation_id)
+
+        if existing_lock:
+            return existing_lock
+
+        return self.create_lock(conversation_id)
+
+    def is_someone_waiting(self, conversation_id: Text) -> bool:
+        """Return whether someone is waiting for lock associated with
+        `conversation_id`."""
+
+        lock = self.get_lock(conversation_id)
+        if lock:
+            return lock.is_someone_waiting()
+
+        return False
+
+    def finish_serving(self, conversation_id: Text, ticket_number: int) -> None:
+        """Finish serving ticket with `ticket_number` for `conversation_id`.
+
+        Removes ticket from lock and saves lock.
+        """
+
+        lock = self.get_lock(conversation_id)
+        if lock:
+            lock.remove_ticket_for_ticket_number(ticket_number)
+            self.save_lock(lock)
+
+    def cleanup(self, conversation_id: Text, ticket_number: int) -> None:
+        """Remove lock for `conversation_id` if no one is waiting."""
+        self.finish_serving(conversation_id, ticket_number)
+        if not self.is_someone_waiting(conversation_id):
+            self.delete_lock(conversation_id)
+            logger.debug(
+                "Deleted lock for conversation '{}' (unused)".format(conversation_id)
+            )
+
+
+class RedisLockStore(LockStore):
+    """Redis store for ticket locks."""
+
+    def __init__(
+        self,
+        host: Text = "localhost",
+        port: int = 6379,
+        db: int = 1,
+        password: Optional[Text] = None,
+        lifetime: int = 60,
+    ) -> None:
+        import redis
+
+        self.red = redis.StrictRedis(
+            host=host, port=int(port), db=int(db), password=password
+        )
+        super().__init__(lifetime)
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        serialised_lock = self.red.get(conversation_id)
+        if serialised_lock:
+            return TicketLock.from_dict(json.loads(serialised_lock))
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        self.red.delete(conversation_id)
+
+    def save_lock(self, lock: TicketLock) -> None:
+        self.red.set(lock.conversation_id, lock.dumps())
+
+
+class InMemoryLockStore(LockStore):
+    """In-memory store for ticket locks."""
+
+    def __init__(self, lifetime: int = 60) -> None:
+        self.conversation_locks = {}  # type: Dict[Text, TicketLock]
+        super().__init__(lifetime)
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        return self.conversation_locks.get(conversation_id)
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        if conversation_id in self.conversation_locks:
+            del self.conversation_locks[conversation_id]
```

Comment:
I think I would do a `try except` thing cause it would be faster than checking whether it's in the locks and then deleting it.